File: src/main/java/player/HumanPlayer.java
package player;

import board.ChessGameInterface;
import helper.Move;
import helper.Position;
import organisation.ChessGameUI;
import organisation.TableInterface;

import javax.swing.*;

/**
 * @author stephan
 */
public class HumanPlayer
        implements HumanPlayerInterface {
    private Position from;
    private ChessGameInterface game;
    private ChessGameUI ui;
    private TableInterface table;
    private boolean isWhitePlayer;
    private boolean isMyTurn;

    public HumanPlayer(TableInterface table,
                       boolean isWhitePlayer,
                       ChessGameUI ui,
                       ChessGameInterface game) {
        this.table = table;
        this.game = game;
        this.ui = ui;
        this.isWhitePlayer = isWhitePlayer;
        isMyTurn = false;
        ui.addPlayer(this);
    }

    public void play() {
        isMyTurn = true;
    }

    public void move(Move move) {
        from = null;
        ui.markPosition(null, true);
        ui.markPosition(null, false);
        isMyTurn = false;
        table.move(move);
    }

    public void mouseMovedOver(Position pos) {
        if (!isMyTurn) return;

        if (from == null) {
            if (game.isSelectable(pos, isWhitePlayer)) {
                ui.markPosition(pos, true);
            } else {
                ui.markPosition(null, true);
            }
        } else {
            if (game.isMoveable(from, pos, isWhitePlayer)) {
                ui.markPosition(pos, false);
            } else {
                ui.markPosition(null, false);
            }
        }
    }

    public void mouseClickedOn(Position pos) {
        if (!isMyTurn) return;

        if (from == null) {
            if (game.isSelectable(pos, isWhitePlayer)) {
                from = pos;
            }
        } else {
            if (game.isMoveable(from, pos, isWhitePlayer)) {
                move(Move.get(from, pos));
            }
        }
    }

    public String askForPawnChange(Position pawnPosition) {
        String[] figs = {"Dame", "Turm", "Läufer", "Springer"};
        String out = (String) JOptionPane.showInputDialog(null,
                "Wähle eine Figur",
                "Bauernumwandlung",
                JOptionPane.QUESTION_MESSAGE,
                null,
                figs,
                "Dame");

        if (out.equals("Dame")) return "Queen";
        if (out.equals("Turm")) return "Rock";
        if (out.equals("Läufer")) return "Bishop";
        return "Knight";
    }

    public void setIsPlaying(boolean isPlaying) {
        isMyTurn = isPlaying && isWhitePlayer;
    }

    public void setColor(boolean isWhite) {
        isWhitePlayer = isWhite;
    }
}

File: module-ui/src/main/kotlin/voidchess/player/human/HumanPlayer.kt
package voidchess.player.human

import voidchess.board.CentralChessGame
import voidchess.common.board.move.Move
import voidchess.common.board.move.MoveResult
import voidchess.common.board.move.PawnPromotion
import voidchess.common.board.move.Position
import voidchess.ui.ChessboardComponent
import voidchess.ui.PosType
import voidchess.ui.TableInterface

import javax.swing.*


class HumanPlayer(
        private val table: TableInterface,
        private var isWhitePlayer: Boolean,
        private val ui: ChessboardComponent,
        private val game: CentralChessGame
) : HumanPlayerInterface {
    private var mouseHoverWhileInactivePos: Position? = null
    private var from: Position? = null
    private var isMyTurn: Boolean = false

    init {
        ui.setPlayer(this)
    }

    override fun play() {
        isMyTurn = true
        mouseMovedOver(mouseHoverWhileInactivePos)
    }

    fun move(move: Move) {
        dropMarkedPositions()
        isMyTurn = false
        table.move(move)
    }

    override fun mouseMovedOver(pos: Position?) {
        if (!isMyTurn) {
            mouseHoverWhileInactivePos = pos
            return
        }

        val lockedFrom = from
        if( pos==null) {
            ui.unmarkPosition(if(lockedFrom==null) PosType.HOVER_FROM else PosType.HOVER_TO)
            return
        }

        if (lockedFrom == null) {
            ui.unmarkPosition(PosType.HOVER_FROM)
            if (game.isSelectable(pos)) {
                ui.markPosition(pos, PosType.HOVER_FROM)
            }
        } else {
            ui.unmarkPosition(PosType.HOVER_TO)
            if (game.isMovable(lockedFrom, pos)) {
                ui.markPosition(pos, PosType.HOVER_TO)
            }
        }
    }

    override fun mouseClickedOn(pos: Position) {
        if (!isMyTurn) return

        val lockedFrom = from
        if (lockedFrom == null) {
            if (game.isSelectable(pos)) {
                from = pos
                ui.markPosition(pos, PosType.SELECT_FROM)
            }
        } else {
            if (game.isMovable(lockedFrom, pos)) {
                move(Move[lockedFrom, pos])
            }
        }
    }

    override fun askForPawnPromotionType(pawnPosition: Position): PawnPromotion {
        val figs = arrayOf("Queen", "Rook", "Bishop", "Knight")
        val out = JOptionPane.showInputDialog(null,
                "Promote pawn to what type?",
                "pawn promotion",
                JOptionPane.QUESTION_MESSAGE, null,
                figs,
                "Queen") as String

        return when (out) {
            "Queen" -> PawnPromotion.QUEEN
            "Rook" -> PawnPromotion.ROOK
            "Knight" -> PawnPromotion.KNIGHT
            else -> PawnPromotion.BISHOP
        }
    }

    override fun gameStarts() {
        isMyTurn = isWhitePlayer
    }

    override fun gameEnds(endoption: MoveResult, lastMoveByWhite: Boolean) {
        isMyTurn = false
        dropMarkedPositions()
    }

    override fun gaveCheck() {
    }

    private fun dropMarkedPositions() {
        from = null
        ui.unmarkPosition(PosType.HOVER_FROM)
        ui.unmarkPosition(PosType.SELECT_FROM)
        ui.unmarkPosition(PosType.HOVER_TO)
    }

    override fun setColor(isWhite: Boolean) {
        isWhitePlayer = isWhite
        ui.setViewPoint(isWhite)
    }
}